src/unix/signal.c

```diff
@@ -360,11 +364,10 @@ void deliver_signal_to_process(process p, struct siginfo *info)
         /*  Attempt to deliver via signalfd notify. If the
             thread becomes runnable, we're done.  */
         signalfd_dispatch(can_wake, sigword);
-        if (thread_is_runnable(can_wake))
-            return;
         sig_debug("attempting to interrupt thread %d\n", can_wake->tid);
         if (!thread_attempt_interrupt(can_wake))
             sig_debug("failed to interrupt\n");
```

Comment:
This is only a debug message, so it should be no issue, but it seems that a thread could possibly wake up between the point that we select it above (being in interruptible sleep) and attempt to interrupt it here - thus making a false return valid here. For that matter, it may be necessary to place a thread_lock / unlock in thread_in_interruptible_sleep around the t->blocked_on checks.

Reply:
thread_in_interruptible_sleep() is called by code (like thread_attempt_interrupt) with the thread lock already held, so we can't place thread_lock / unlock in this function.
Yes, it is possible that a thread wakes up between the point it's selected and the attempt to interrupt it; but this is the case even with the kernel lock, because a thread can be woken up by a bhqueue thunk running without the kernel lock. If that happens, the signal will be delivered at a later point, most likely the next time a thread (possibly a different one) is about to be scheduled again.
Anyway, I re-added the call to thread_is_runnable() so that if the thread is woken up via signalfd we don't attempt to interrupt it.